### PR TITLE
Update the installer for more extension types

### PR DESCRIPTION
**Pull request by jsuchome**
- make the installation scripts executable
- allow empty namespaces
- create namespace when it does not exist
- added support for kustomize dir

TODO: we're missing this condition from current knative installation script:
`kubectl wait --for=condition=available --timeout=600s deployment -l networking.knative.dev/ingress-provider=istio -n knative-serving`
I've introduced `waitfor` option, but it's currently very simple. There could be more complicated definition, like
```
install:
  - type: manifest
    location: net-istio.yaml
    namespace:
    waitfor:
      - type: deployment
        selector:  networking.knative.dev/ingress-provider=istio
        namespace: knative-serving
        status: available
      - type: pod
        selector: all
        status: running
        namespace: knative-serving
```
but isn't it overkill?
It could be simpler to actually add a `script` type as a next step, that would do the waiting...

UPDATE: I actually added extra script to solve the wait problem in https://github.com/fuseml/extensions/pull/10. Later, we can get rid of it

### CODE REVIEW

**Comment by flaviodsr** [comment]
I think think it would be good to extend the `waitfor` as you described, but instead of having multiple params you could simple have something like, for example, `wait: "--for=delete pod/busybox1 --timeout=60s"` which would be easier to implement and allow it to be used with all the options `kubectl` offers.

**Comment by jsuchome** [comment]
> I think think it would be good to extend the `waitfor` as you described, but instead of having multiple params you could simple have something like, for example, `wait: "--for=delete pod/busybox1 --timeout=60s"` which would be easier to implement and allow it to be used with all the options `kubectl` offers.

Hm, right, this would be easy to implement... it is too much `kubectl` specific, but I'm not sure if it is a problem...

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -118,7 +118,7 @@ jobs:
       - name: Install FuseML
         run: |
           ./dist/fuseml-installer version
-          make fuseml-install-with-mlflow
+          make fuseml-install-with-extensions
 
       - name: Install KFServing
```
Comment (flaviodsr) [inline .github/workflows/ci.yml:123]: At this point you can also delete this step.

Comment (jsuchome) [reply .github/workflows/ci.yml:123]: Right, I missed that one

**Comment by stefannica** [comment]
I think you had the right idea with adding a `waitfor` option. Sure, you could do that with a bash script, but it makes things much simpler when you're just fetching some kustomize dir from upstream and you need to wait for services to come up, no extra bits required :)

**Comment by stefannica** [comment]
> I think think it would be good to extend the `waitfor` as you described, but instead of having multiple params you could simple have something like, for example, `wait: "--for=delete pod/busybox1 --timeout=60s"` which would be easier to implement and allow it to be used with all the options `kubectl` offers.

If you do it like this, why just settle for kubectl commands, better to add support to specify entire shell scripts inline in the yaml file.

**Comment by jsuchome** [comment]
> > I think think it would be good to extend the `waitfor` as you described, but instead of having multiple params you could simple have something like, for example, `wait: "--for=delete pod/busybox1 --timeout=60s"` which would be easier to implement and allow it to be used with all the options `kubectl` offers.
> 
> If you do it like this, why just settle for kubectl commands, better to add support to specify entire shell scripts inline in the yaml file.

Well, that's what we have now, basically with the scripts treated as extra steps.

Do you mean you prefer my original suggestion written on top?

**Review by stefannica — DISMISSED** [review]
Great PR, I especially liked the `kustomize` and `waitfor` addition, it's the type of thing that will make it really easy to support more extensions without incurring tech debt. Thank you !

**Comment by stefannica** [comment]
> > > I think think it would be good to extend the `waitfor` as you described, but instead of having multiple params you could simple have something like, for example, `wait: "--for=delete pod/busybox1 --timeout=60s"` which would be easier to implement and allow it to be used with all the options `kubectl` offers.
> > 
> > 
> > If you do it like this, why just settle for kubectl commands, better to add support to specify entire shell scripts inline in the yaml file.
> 
> Well, that's what we have now, basically with the scripts treated as extra steps.
> 
> Do you mean you prefer my original suggestion written on top?

Yes, but I see you already implemented it as an option, which is great.

**Comment by stefannica** [comment]
> > > 
> > > If you do it like this, why just settle for kubectl commands, better to add support to specify entire shell scripts inline in the yaml file.
> > 
> > 
> > Well, that's what we have now, basically with the scripts treated as extra steps.

Wait, that's not what I meant. I meant _inline_ shell scripts, instead of having to point to another file alongside the YAML descriptor.

> > Do you mean you prefer my original suggestion written on top?
> 
> Yes, but I see you already implemented it as an option, which is great.

Oh, I see, you only implemented the `wait pods` option, but otherwise `wait` doesn't do anything.

**Review by stefannica — APPROVED** [review]

**Pull request merged**